CombinatorialGames/Game/Computability/FGame.lean

```diff
@@ -0,0 +1,418 @@
+    Relator.BiTotal (x.moveLeft · ≡ y.moveLeft ·) ∧ Relator.BiTotal (x.moveRight · ≡ y.moveRight ·)
+  | mk .., mk .. => Iff.rfl
+
+@[refl]
+protected theorem Identical.refl (x) : x ≡ x :=
+  x.recOn fun _ _ _ _ IHL IHR ↦ ⟨Relator.BiTotal.refl IHL, Relator.BiTotal.refl IHR⟩
+
+@[symm]
+protected theorem Identical.symm : ∀ {x y}, x ≡ y → y ≡ x
+  | mk .., mk .., ⟨hL, hR⟩ => ⟨hL.symm fun _ _ h ↦ h.symm, hR.symm fun _ _ h ↦ h.symm⟩
+
+@[trans]
+protected theorem Identical.trans : ∀ {x y z}, x ≡ y → y ≡ z → x ≡ z
+  | mk .., mk .., mk .., ⟨hL₁, hR₁⟩, ⟨hL₂, hR₂⟩ =>
+    ⟨hL₁.trans (fun _ _ _ h₁ ↦ h₁.trans) hL₂, hR₁.trans (fun _ _ _ h₁ ↦ h₁.trans) hR₂⟩
+
+/-- `Identical` as a `Setoid`. -/
+def identicalSetoid : Setoid SGame :=
+  ⟨Identical, .refl, .symm, .trans⟩
+
+/-- If `x ≡ y`, then a left move of `x` is identical to some left move of `y`. -/
+theorem Identical.moveLeft : ∀ {x y}, x ≡ y → ∀ i, ∃ j, x.moveLeft i ≡ y.moveLeft j
+  | mk .., mk .., ⟨hl, _⟩ => hl.1
+
+/-- If `x ≡ y`, then a left move of `y` is identical to some left move of `x`. -/
+theorem Identical.moveLeft_symm : ∀ {x y}, x ≡ y → ∀ i, ∃ j, x.moveLeft j ≡ y.moveLeft i
+  | mk .., mk .., ⟨hl, _⟩ => hl.2
+
+/-- If `x ≡ y`, then a right move of `x` is identical to some right move of `y`. -/
+theorem Identical.moveRight : ∀ {x y}, x ≡ y → ∀ i, ∃ j, x.moveRight i ≡ y.moveRight j
+  | mk .., mk .., ⟨_, hr⟩ => hr.1
+
+/-- If `x ≡ y`, then a right move of `y` is identical to some right move of `x`. -/
+theorem Identical.moveRight_symm : ∀ {x y}, x ≡ y → ∀ i, ∃ j, x.moveRight j ≡ y.moveRight i
+  | mk .., mk .., ⟨_, hr⟩ => hr.2
+
+instance Identical.instDecidable (a b) : Decidable (a ≡ b) := by
+  have : DecidableRel (a.moveLeft · ≡ b.moveLeft ·) := fun c d ↦ Identical.instDecidable ..
+  have : DecidableRel (a.moveLeft · ≡ b.moveRight ·) := fun c d ↦ Identical.instDecidable ..
+  have : DecidableRel (a.moveRight · ≡ b.moveLeft ·) := fun c d ↦ Identical.instDecidable ..
+  have : DecidableRel (a.moveRight · ≡ b.moveRight ·) := fun c d ↦ Identical.instDecidable ..
+  rw [identical_iff]
+  unfold Relator.BiTotal Relator.LeftTotal Relator.RightTotal
+  infer_instance
+termination_by (a, b)
+decreasing_by sgame_wf
+
+/-- Converts an SGame to a PGame -/
+def toPGame (x : SGame) : PGame := .mk _ _ (toPGame <| x.moveLeft ·) (toPGame <| x.moveRight ·)
+termination_by x
+decreasing_by sgame_wf
+
+@[simp]
+theorem toPGame_leftMoves {x : SGame} : x.toPGame.LeftMoves = Fin x.LeftMoves := by
+  unfold toPGame; rfl
+
+@[simp]
+theorem toPGame_rightMoves {x : SGame} : x.toPGame.RightMoves = Fin x.RightMoves := by
+  unfold toPGame; rfl
+
+theorem toPGame_identical {x y : SGame} (h : x ≡ y) : (toPGame x).Identical (toPGame y) := by
+  rw [identical_iff] at h
+  rw [PGame.identical_iff]
+  constructor
+  · constructor
+    · obtain ⟨⟨h, -⟩, -⟩ := h
+      intro y
+      obtain ⟨b, hb⟩ := h (toPGame_leftMoves.mp y)
+      use toPGame_leftMoves.mpr b
+      dsimp
+      sorry
+    · sorry
+  · sorry
```

Comment:
You can avoid those horrible casts by rewriting the definition of `toPGame` from the very start:
```suggestion
  rw [toPGame, toPGame, PGame.identical_iff]
  refine ⟨⟨?_, ?_⟩, ⟨?_, ?_⟩⟩ <;> intro k
  on_goal 1 => obtain ⟨b, hb⟩ := h.1.1 k
  on_goal 2 => obtain ⟨b, hb⟩ := h.1.2 k
  on_goal 3 => obtain ⟨b, hb⟩ := h.2.1 k
  on_goal 4 => obtain ⟨b, hb⟩ := h.2.2 k
  all_goals exact ⟨b, toPGame_identical hb⟩
termination_by (x, y)
decreasing_by sgame_wf
```